src/main/java/org/junitpioneer/jupiter/CartesianEnumArgumentsProvider.java

```diff
@@ -44,14 +52,50 @@ public void accept(CartesianEnumSource enumSource) {
 			if (uniqueNames.size() != declaredConstantNames.length)
 				throw new PreconditionViolationException("Duplicate enum constant name(s) found in " + enumSource);
 
-			mode.validate(enumSource, uniqueNames);
-			this.constants.removeIf(constant -> !mode.select(constant, uniqueNames));
+			mode.validate(enumSource, constants, uniqueNames);
+			constants.removeIf(constant -> !mode.select(constant, uniqueNames));
 		}
+		return constants.stream().map(Arguments::of);
 	}
 
-	@Override
-	public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
-		return constants.stream().map(Arguments::of);
+	private <E extends Enum<E>> Set<? extends E> getEnumConstants(ExtensionContext context) {
+		Class<E> enumClass = determineEnumClass(context);
+		return EnumSet.allOf(enumClass);
+	}
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	private <E extends Enum<E>> Class<E> determineEnumClass(ExtensionContext context) {
+		Class enumClass = enumSource.value();
+		if (enumClass.equals(NullEnum.class)) {
+			Method method = context.getRequiredTestMethod();
+			Class<?>[] parameterTypes = method.getParameterTypes();
+
+			if (parameterTypes.length <= 0)
+				throw new PreconditionViolationException(
+					"Test method must declare at least one parameter: " + method.toGenericString());
+
+			Class<?> parameterType = parameterTypes[determineParameterTypeIndex(method)];
+
+			if (!Enum.class.isAssignableFrom(parameterType))
+				throw new PreconditionViolationException(String
+						.format(
+							"Parameter of type %s must reference an Enum type (alternatively, use the annotation's 'value' attribute to specify the type explicitly): %s",
+							parameterType, method.toGenericString()));
+
+			enumClass = parameterType;
+		}
+		return enumClass;
+	}
+
+	private int determineParameterTypeIndex(Method method) {
+		List<Annotation> argumentSources = PioneerAnnotationUtils
+				.findAnnotatedAnnotations(method, ArgumentsSource.class);
+
+		return IntStream
```

Comment:
While I love a good stream as much as the next guy, this is probably a bit too much for an operation that is essentially just `List::indexOf`.

Reply:
This part didn't start as a `List` and later I didn't realize I could improve it, thanks for spotting it.

Reply:
Actually, `indexOf` is not a good replacement in this case because it relies on `equals` while I really want reference comparison here. The use case is when there are several `@CartesianEnumSource`s, all with omitted types (when they don't hit #447, of course)

Reply:
Fair. I had to look up how [Object#equals](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)) differs from [Annotation#equals](https://docs.oracle.com/javase/7/docs/api/java/lang/annotation/Annotation.html#equals(java.lang.Object)). I learned something new! 🎉